[visionOS] Remote Effects don't work inside of `pointer-events: none` scrollers

etiennesegonzac opened the pull request:
#### 0fed4da451372d464b08bd8a0545e8a9024bb49c
<pre>
[visionOS] Remote Effects don&apos;t work inside of `pointer-events: none` scrollers
<a href="https://bugs.webkit.org/show_bug.cgi?id=262307">https://bugs.webkit.org/show_bug.cgi?id=262307</a>
&lt;<a href="https://rdar.apple.com/115776753">rdar://115776753</a>&gt;

Reviewed by NOBODY (OOPS!).

Remote Effects use CA-based hit-testing.
So when we set `userInteractionEnabled = NO` on a `pointer-events: none`
ScrollView, we also prevent potential `pointer-events: auto` children
from hit-testing.

This patch introduces a new `_wk_userInteractive` flag on
WKChildScrollViews that disables the scroll container without blocking
hit-testing of its children. Effectively taking the view out of
WKHitTesting without impacting its layer&apos;s `allowsHitTesting` property.

* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTreePropertyApplier.mm:
(WebKit::RemoteLayerTreePropertyApplier::applyPropertiesToUIView):
Use `_wk_userInteractive` instead of `userInteractionEnabled` for
WKChildScrollViews.

* Source/WebKit/UIProcess/RemoteLayerTree/ios/RemoteLayerTreeViews.h:
* Source/WebKit/UIProcess/RemoteLayerTree/ios/RemoteLayerTreeViews.mm:
(-[WKChildScrollView initWithFrame:]):
Introduce the new `_wk_userInteractive` property defaulting to YES.
(WebKit::collectDescendantViewsAtPoint):
(WebKit::collectDescendantViewsInRect):
Remove the FIXME since we use `userInteractionEnabled` for
`WKNativelyInteractible` views.
Skip WKChildScrollViews marked as `_wk_userInteractive = NO`.

* Source/WebKit/UIProcess/RemoteLayerTree/ios/ScrollingTreeScrollingNodeDelegateIOS.h:
Clean-up, remove unused member variable.

* Source/WebKit/UIProcess/API/Cocoa/WKWebViewTesting.mm:
(dumpCALayer):
Add `allowsHitTesting` to the dump format when Interaction Regions are
turned on.

* LayoutTests/interaction-region/scroll-view-hit-testing-expected.txt: Added.
* LayoutTests/interaction-region/scroll-view-hit-testing.html: Added.
Add a test covering the change, based on
`touch-scroll-pointer-events-none.html`.
* LayoutTests/interaction-region/interaction-layers-culling-expected.txt:
* LayoutTests/interaction-region/layer-tree-expected.txt:
Update expectations with the new format.
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0fed4da451372d464b08bd8a0545e8a9024bb49c

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/32556 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/11304 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/34404 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/35122 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/29431 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/13653 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/8498 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/28960 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/32990 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/9526 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/29094 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/8304 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/8440 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/29036 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/36458 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/29590 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/29476 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/34562 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/8568 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/6524 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/32421 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/10241 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/28780 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/9180 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/9158 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/rendering/RenderLayerScrollableArea.h:
```diff
@@ -225,6 +225,9 @@ class RenderLayerScrollableArea final : public ScrollableArea {
     void updateScrollbarsAfterStyleChange(const RenderStyle* oldStyle);
     void updateScrollbarsAfterLayout();
 
+    ScrollbarMode horizontalScrollbarMode() const final;
```
I don't think `pointer-events:none` should affect the scrollbar mode (it doesn't affect how scrollbars render, only whether the user can interact).

etiennesegonzac replied:
it was a bit fragile having completely different codepaths calling into [setScrollEnabled](https://github.com/WebKit/WebKit/blob/9ea306f30d73877433b5545e71f77fd1e64b1d9d/Source/WebKit/UIProcess/RemoteLayerTree/ios/ScrollingTreeScrollingNodeDelegateIOS.mm#L305) anyway.

Trying a simpler approach: a custom property on `WKChildScrollView`.

smfr commented on LayoutTests/interaction-region/scroll-view-hit-testing.html:
```diff
@@ -0,0 +1,67 @@
+        height: 300px;
+        width: 300px;
+        overflow: scroll;
+        -webkit-overflow-scrolling: touch;
```
This doesn't do anything.

pxlcoder commented on Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTreePropertyApplier.mm:
```diff
@@ -409,8 +409,13 @@ static void updateCustomAppearance(CALayer *layer, GraphicsLayer::CustomAppearan
+    if (properties.changedProperties.containsAny({ LayerChange::ContentsHiddenChanged, LayerChange::UserInteractionEnabledChanged })) {
+        bool userInteractive = !properties.contentsHidden && properties.userInteractionEnabled;
+        if ([view isKindOfClass:[WKChildScrollView class]])
+            [(WKChildScrollView *)view setScrollEnabled:userInteractive];
```
```suggestion
        if (auto scrollView = dynamic_objc_cast<WKChildScrollView>(view))
            [scrollView setScrollEnabled:userInteractive];
```

Similar comment in two other places.

etiennesegonzac replied:
This one is fixed (and consistent with the rest of the file), but `RemoteLayerTreeViews.mm` is _full_ of `isKindOfClass`, are we progressively migrating file by file?

pxlcoder replied:
I don't think there's a concerted effort to migrate – it's just preferred since it removes the redundant class name. Also, there are probably many people working on t WebKit that aren't aware of `dynamic_objc_cast`.

I'd be okay if you left `RemoteLayerTreeViews.mm` as-is though.

etiennesegonzac commented on Source/WebCore/page/scrolling/ScrollingTreeScrollingNode.cpp:
```diff
@@ -160,6 +160,11 @@ bool ScrollingTreeScrollingNode::canHandleWheelEvent(const PlatformWheelEvent& w
+{
+    return m_scrollableAreaParameters.allowsHorizontalScrolling || m_scrollableAreaParameters.allowsVerticalScrolling;
+}
+
```
`check-webkit-styles` told me not to inline it because of the class-wide `WEBCORE_EXPORT` but the class is already full on inline methods :)

etiennesegonzac committed:
[visionOS] Remote Effects don't work inside of `pointer-events: none` scrollers
https://bugs.webkit.org/show_bug.cgi?id=262307
<rdar://115776753>

Reviewed by NOBODY (OOPS!).

Remote Effects use CA-based hit-testing.
So when we set `userInteractionEnabled = NO` on a `pointer-events: none`
ScrollView, we also prevent potential `pointer-events: auto` children
from hit-testing.

This patch introduces a new `_wk_userInteractive` flag on
WKChildScrollViews that disables the scroll container without blocking
hit-testing of its children. Effectively taking the view out of
WKHitTesting without impacting its layer's `allowsHitTesting` property.

* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTreePropertyApplier.mm:
(WebKit::RemoteLayerTreePropertyApplier::applyPropertiesToUIView):
Use `_wk_userInteractive` instead of `userInteractionEnabled` for
WKChildScrollViews.

* Source/WebKit/UIProcess/RemoteLayerTree/ios/RemoteLayerTreeViews.h:
* Source/WebKit/UIProcess/RemoteLayerTree/ios/RemoteLayerTreeViews.mm:
(-[WKChildScrollView initWithFrame:]):
Introduce the new `_wk_userInteractive` property defaulting to YES.
(WebKit::collectDescendantViewsAtPoint):
(WebKit::collectDescendantViewsInRect):
Remove the FIXME since we use `userInteractionEnabled` for
`WKNativelyInteractible` views.
Skip WKChildScrollViews marked as `_wk_userInteractive = NO`.

* Source/WebKit/UIProcess/RemoteLayerTree/ios/ScrollingTreeScrollingNodeDelegateIOS.h:
Clean-up, remove unused member variable.

* Source/WebKit/UIProcess/API/Cocoa/WKWebViewTesting.mm:
(dumpCALayer):
Add `allowsHitTesting` to the dump format when Interaction Regions are
turned on.

* LayoutTests/interaction-region/scroll-view-hit-testing-expected.txt: Added.
* LayoutTests/interaction-region/scroll-view-hit-testing.html: Added.
Add a test covering the change, based on
`touch-scroll-pointer-events-none.html`.
* LayoutTests/interaction-region/interaction-layers-culling-expected.txt:
* LayoutTests/interaction-region/layer-tree-expected.txt:
Update expectations with the new format.
